### DISCUSSION TITLE
How is the user for a push to a GitHub fork origin chosen?

### DISCUSSION THREAD

**Discussion by xyzzy-plugh-plover** [community/community#36623] [Other Feature Feedback, Questions, & Ideas] [answered]
How is the user for a push to a GitHub fork origin chosen?

I have an existing user that uses an SSH key, but my new user uses a GPG key.

My origin fork is for the GPG user, and all commits & branches in my local clone are for the GPG user.

When I try to push a new branch to my origin fork, however, it fails because it's using the SSH user.

How can I push using the GPG user?

I am running macOS 12.6, git 2.38.1 & gh 2.18.0.

**Comment by airtower-luna** [comment, marked as answer]
What exactly do you mean by "GPG user"? You can't authorize a push to GitHub via GPG, only SSH (with a key) or HTTPS (with a token).

You can make SSH use a private key that's in your GPG keyring, but from the server (GitHub) point of view that's still SSH auth, so you'd need to add the matching private key to the account as an SSH key.

**Reply by xyzzy-plugh-plover** [reply]
I added 2 SSH keys to my new account (the one I previously called my GPG account), one for signing & one for authentication.

I have them locally in my `~/.ssh`, added them to my SSH agent, and added the authentication key to my `~/.ssh/config`, which now looks like:

```
Host *
 AddKeysToAgent yes
 UseKeychain yes
 IdentityFile ~/.ssh/id_ed25519
Host *.github.com
 AddKeysToAgent yes
 UseKeychain yes
 IdentityFile ~/.ssh/id_xyzzy-plugh-plover_thrawn_ed25519
```

When I push to my fork, it still tries to push from my old user (which I previously called my SSH user) instead of using the new user.

How can I push using the correct user, preferably based on a per clone basis? i.e. I want to push using my old account for clones that use that one (for signed commits & for the origin on github), while using the new account for clones that use that one (for signed commits & for the origin on github).

**Reply by airtower-luna** [reply]
I spot two problems there. For the first one, I'll quote the [`ssh_config` manual page](https://manpages.debian.org/bullseye/openssh-client/ssh_config.5.en.html):

> Since the first obtained value for each parameter is used, more host-specific declarations should be given near the beginning of the file, and general defaults at the end.

The `Host *` section applies to every host, so its `IdentityFile` setting will be used for every host. Move it to the end of the file to change that.

The `Host *.github.com` line won't match `github.com`. Depending on how you configure the SSH URLs you could either remove the `*.`, or add `github.com` (after a space).

The user is simply detected by the key used for authentication. If you want *multiple* configs for `github.com`, you can define additional hosts for them, and use the [`Hostname` setting](https://manpages.debian.org/bullseye/openssh-client/ssh_config.5.en.html#Hostname) to point them at the right host. For example:

```
Host gh2
  Hostname github.com
  User git
  AddKeysToAgent yes
  IdentityFile ~/.ssh/another_id
```

This tells SSH to connect to `github.com` with another ID file and the user `git` when you ask it to connect to `gh2`. So you can just replace `git@github.com` in the SSH URL with `gh2` to make it use that identity.

**Reply by xyzzy-plugh-plover** [reply]
Thanks

**Comment by ossanna16** [comment]
Hey there @xyzzy-plugh-plover and thank you for asking a good question! I'm glad you found our community 🙂
